### PR TITLE
Enables builds with Compute Library backend for oneDNN

### PR DESCRIPTION
**Pull request by nSircombe**
Since v1.7, oneDNN (MKL-DNN) has supported the use of Compute Library
for the Arm architeture to provide optimised convolution primitives
on AArch64.

This change enables the use of Compute Library in the PyTorch build.
Following the approach used to enable the use of CBLAS in MKLDNN,
It is enabled by setting the env vars USE_MKLDNN and USE_MKLDNN_ACL.
The location of the Compute Library build must be set useing `ACL_ROOT_DIR`.

This is an extension of the work in https://github.com/pytorch/pytorch/pull/50400
which added support for the oneDNN/MKL-DNN backend on AArch64.

_Note: this assumes that Compute Library has been built and installed at
ACL_ROOT_DIR. Compute library can be downloaded here:
`https://github.com/ARM-software/ComputeLibrary`_

Fixes #{issue number}

### CODE REVIEW

**Comment by nSircombe** [comment]
Hi @malfet, @jbschlosser, @walterddr 

It looks like some of the Circle CI stages are failing - the logs appear to indicate that it can't find the new `ComputeLibrary.cmake` file, but I don't see why. I've been unable to replicate this failure on my own machines, am I missing something?

...the file does appear to be there, since the linter initially caught a minor issue with in `ComputeLibrary.cmake` initially.

File: CMakeLists.txt [hunk 1]
```diff
@@ -264,6 +264,9 @@ option(USE_ZSTD "Use ZSTD" OFF)
 cmake_dependent_option(
   USE_MKLDNN "Use MKLDNN. Only available on x86, x86_64, and AArch64." "${CPU_INTEL}"
   "CPU_INTEL OR CPU_AARCH64" OFF)
+cmake_dependent_option(
+  USE_MKLDNN_ACL "Use Compute Library for the Arm architecture." OFF
+  "USE_MKLDNN AND CPU_AARCH64 AND USE_MKLDNN_ACL" OFF)
```
Comment (malfet) [inline CMakeLists.txt:269]: Why usemkldnnacl definition depends on itself?

File: cmake/public/ComputeLibrary.cmake [hunk 2]
```diff
@@ -0,0 +1,41 @@
+# Build with Compute Library backend for the Arm architecture
+# Note: Compute Library is available from: https://github.com/ARM-software/ComputeLibrary
+#   and must be built separately. The location of the Compute Library build
+#   must be set with the env var ACL_ROOT_DIR. This path will be checked later
+#   as part of FindACL.cmake in oneDNN.
+
+IF(NOT CPU_AARCH64)
```
Comment (malfet) [inline cmake/public/ComputeLibrary.cmake:7]: Please use lowercase spelling of CMake commands (for consistency with rest of files in the folder)

File: cmake/public/ComputeLibrary.cmake [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  IF(${ACL_VERSION} VERSION_EQUAL "0.0")
+    # Unreleased ACL versions come with version string "v0.0-unreleased", and may not be compatible with oneDNN.
+    # It is recommended to use the latest release of ACL.
+    MESSAGE(WARNING "Build may fail: Using unreleased ACL version (minimum required is ${ACL_MINIMUM_VERSION})")
+  ELSEIF(${ACL_VERSION} VERSION_LESS ${ACL_MINIMUM_VERSION})
+    MESSAGE(FATAL_ERROR "Detected ACL version ${ACL_VERSION}, but minimum required is ${ACL_MINIMUM_VERSION}")
+  ENDIF()
+ENDIF()
+
+SET(CMAKE_CXX_STANDARD 14)
```
Comment (malfet) [inline cmake/public/ComputeLibrary.cmake:40]: Please do not define C++ standard for the entire project here.

**Review by malfet — COMMENTED** [review]
Most likely error happens because new .cmake file is not installed.

PS one can workaround the problem by including the file only if CPU_ARM64 is defined

File: CMakeLists.txt (hunk 1 above)
Comment (nSircombe) [reply CMakeLists.txt]: Ah, just a mistake - that's gone.

File: cmake/public/ComputeLibrary.cmake (hunk 2 above)
Comment (nSircombe) [reply cmake/public/ComputeLibrary.cmake]: Done.

File: cmake/public/ComputeLibrary.cmake (hunk 3 above)
Comment (nSircombe) [reply cmake/public/ComputeLibrary.cmake]: As it turns out this isn't needed here - I've taken it out.

**Comment by nSircombe** [comment]
Thanks for taking a look @malfet & @walterddr. I have updated in line with your suggestions.

**Review by malfet — APPROVED** [review]

**Comment by nSircombe** [comment]
Just looking at the failing checks:
- The  `Lint / flake8-py3` failure does not appear to be associated with these changes. 
- Many CircleCI failures are at the 'Merge target branch' stage.
- Two CIrcleCI checks fail in `test_py_tensors_multi_async_call` but it's not clear that these are associated with these changes either.

@malfet, @walterddr, are there any issues here that I need to address?

**Comment by nSircombe** [comment]
Hi @walterddr, @malfet,

I'd hoped this might make it into the 1.9 release, but it's still waiting on a 2nd review. Will it be possible to get this into the next RC now?

**Comment by malfet** [comment]
@nSircombe Can you please rebase PR again 'viable/strict' branch? (I agree that CI failures are unrelated but it's better have all checks green)

**Comment by nSircombe** [comment]
Thanks @malfet, will do.

**Review by walterddr — APPROVED** [review]

**Comment by nSircombe** [comment]
Thanks for all your help with this @malfet, @walterddr. 
Will it be possible to get this into the next 1.9 RC?

**Comment by malfet** [comment]
As source option? Sure, please create cherry-pick PR and reference it in comment to  https://github.com/pytorch/pytorch/issues/58518

**Comment by nSircombe** [comment]
> As source option?

Yep, please.

I've cherry-picked it onto `release/1.9` and raised a PR - https://github.com/pytorch/pytorch/pull/58746.


Cheers.

**Pull request closed without merging**